src/main/java/com/google/sps/data/Journey.java

```diff
@@ -0,0 +1,218 @@
+   * value in the waypoints hashmap, which is the index in the ArrayList.
+   */
+  private ArrayList<Point> getWaypointsList() {
+    ArrayList<Point> waypointsList = new ArrayList<>();
+    waypointsList.addAll(Collections.nCopies(numberOfWaypoints, start));
+    for (Map.Entry<Point, Integer> entry : waypoints.entrySet()) {
+      waypointsList.set(entry.getValue(), entry.getKey());
+    }
+    return waypointsList;
+  }
+
+  /**
+   * Returns the reverse order in which the waypoints must be visited for an optimal journey.
+   */
+  public ArrayList<Point> getOrderedWaypoints() {
+    return orderedWaypoints;
+  }
+
+  /**
+   * Computes for each waypoint i, pointsAfter[i] = bitmask where bits set to 1 represent the
+   * indices of points that must be visited after point i
+   */
+  public ArrayList<Integer> getPointsAfterFromRestricitions() {
+    ArrayList<Integer> pointsAfter = new ArrayList<>(Collections.nCopies(numberOfWaypoints, 0));
+    for (Pair restriction : restrictions) {
+      // restriction.second must be visited after restriction.first
+      int currentBitmask = pointsAfter.get(restriction.first);
+      pointsAfter.set(restriction.first, currentBitmask | (1 << restriction.second));
+    }
+    return pointsAfter;
+  }
+
+  /**
+   * Given bestTime matrix, it computes and returns the optimal order in which points should be
+   * visited starting from the last point.
+   */
+  public ArrayList<Point> getOrderFromMatrix(ArrayList<ArrayList<Pair>> bestTime) {
+    ArrayList<Point> orderedWaypoints = new ArrayList<>();
+    // the last state of the matrix has all waypoints set to 1
+    int currentState =  (1 << numberOfWaypoints) - 1;
+    // get waypointsList to access the keys from waypoints given their values
+    ArrayList<Point> waypointsList = getWaypointsList();
+    int minTime = bestTime.get((1 << numberOfWaypoints)).get(0).first;
+    // currentWaypoint is the one for which bestTime[currentState][currentWaypoint] is minimal
+    int currentWaypoint = 0;
+    for (int waypointIndex = 1; waypointIndex < numberOfWaypoints; ++waypointIndex) {
+      if (bestTime.get(currentState).get(waypointIndex).first < minTime) {
+        minTime = bestTime.get(currentState).get(waypointIndex).first;
+        currentWaypoint = waypointIndex;
+      }
+    }
+    while (currentState != 0) {
+      orderedWaypoints.add(waypointsList.get(currentWaypoint));
+      int prevState = currentState ^ (1 << currentWaypoint);
+      int prevWaypoint = bestTime.get(currentState).get(currentWaypoint).second;
+      currentWaypoint = prevWaypoint;
+      currentState = prevState;
+    }
+    return orderedWaypoints;
+  }
+
+  /**
+   * Using the bestTime matrix computed in findOptimalTimePaths(), it returns true if the minimum
+   * time journey fits in the timeslot and false otherwise
+   */
+  public boolean findJourneyForTimeslot(Date deliveryDay, int startTime, int endTime) {
```

Comment:
I'm not sure I quite understand what's going on here.

This is the only method tested. Yet this method simply returns a boolean. How will this be useful?

Shouldn't this return the actual journey?

Reply:
In `DeliverySystem.java`, when assigning orders to a delivery request, `findJourneyForTimeslot` is needed for knowing if it is possible to find a journey that delivers the assigned orders (the actual journey matters only when it is displayed to the user)